core/base_abstractions/sensor.py

```diff
@@ -158,23 +166,34 @@ def _get_observation_space(self) -> gym.spaces.Tuple:
         only if the expert failed to generate a true expert action. The
         value `num actions in task` should be in `config["nactions"]`
         """
-        return gym.spaces.Tuple(
-            (gym.spaces.Discrete(self.nactions), gym.spaces.Discrete(2))
-        )
+        return su.flatten_space(self.unflattened_observation_space)
 
     def get_observation(
         self, env: EnvType, task: SubTaskType, *args: Any, **kwargs: Any
     ) -> Any:
         # If the task is completed, we needn't (perhaps can't) find the expert
         # action from the (current) terminal state.
         if task.is_done():
-            return np.array([-1, False], dtype=np.int64)
+            return np.array([self.action_space.sample(), False])
+
         action, expert_was_successful = task.query_expert(**self.expert_args)
-        assert isinstance(action, int), (
-            "In expert action sensor, `task.query_expert()` "
-            "did not return an integer action."
+
+        if isinstance(action, int):
+            assert isinstance(self.action_space, gym.spaces.Discrete)
+            unflattened_action = action
+        else:
+            # Assume we receive a gym-flattened numpy action
+            unflattened_action = gyms.unflatten(self.action_space, action)
```

Comment:
Why would the action returned by the `query_expert` function be flattened? I thought the flattening was mostly for internal eyes only?

Reply:
It was convenient for the engine implementation, but it can certainly be sent unflattened (this was initially intended to reuse code with the old implementation where we only had Categorical distributions).